### DISCUSSION TITLE
Programmatic streaming now supported with `for partial in program(stream=True)`

### DISCUSSION THREAD

**Discussion by slundberg** [guidance-ai/guidance#129] [Announcements]
Guidance now supports the convenient streaming of partial results programmatically (before it was just streamed to the notebook for display).

**Here is an example of synchronous streaming**

```python
import guidance
guidance.llm = guidance.llms.Transformers("databricks/dolly-v2-3b", caching=False)

program = guidance("""Generate a list of 5 company names:
{{#geneach 'companies' num_iterations=5~}}
{{@index}}. "{{gen 'this'}}"
{{/geneach}}""")

for p in program(stream=True, silent=True):
    print(p.get("companies", []))
```
<img width="619" alt="image" src="https://github.com/microsoft/guidance/assets/3740613/d2477b8d-2718-4787-a089-d78039908ef2">

**Here is an example of asynchronous streaming** (run in a notebook event loop)
```python
import guidance
guidance.llm = guidance.llms.Transformers("databricks/dolly-v2-3b", caching=False)

program = guidance("""Generate a list of 5 company names:
{{#geneach 'companies' num_iterations=5~}}
{{@index}}. "{{gen 'this'}}"
{{/geneach}}""")

async for p in program(stream=True, async_mode=True, silent=True):
    print(p.get("companies", []))
```
<img width="619" alt="image" src="https://github.com/microsoft/guidance/assets/3740613/d2477b8d-2718-4787-a089-d78039908ef2">

**Here is an example of tracking just the new content**
```python
import guidance
guidance.llm = guidance.llms.Transformers("databricks/dolly-v2-3b", caching=False)

program = guidance("""Generate a list of 5 company names:
{{gen 'companies' stop="\\n\\n"}}""")

pos = 0
for p in program(stream=True, silent=True):
    val = p.get("companies", "")
    print(val[pos:], end="", flush=True)
    pos = len(val)
```
<img width="487" alt="image" src="https://github.com/microsoft/guidance/assets/3740613/58b54a18-2e5d-4b7c-8315-dcbbf884931d">

**How does it work?**
It returns a reference to the partially completed program object at the same refresh rate used for notebook display streaming. If you want to get just the difference of what might be new, you can track the current position in whatever variable of interest yourself during the loop.

**Caveats**
Since this is a new feature I expect we will need to update some of the library functions to set variable values more frequently so they stream nicer. If you run into a scenario like this post an issue (or a PR!).

**Comment by fracorallo** [comment]
It works perfectly, thanks @slundberg !

Do you also plan to support the update of the internal program conversation "history" ( program.text ), as it happens without streaming?
e.g.:
```python
program = program(stream=False, silent=True)  # this updates the conversation history internally
```

**Comment by AtmehEsraa** [comment]
AttributeError: module 'guidance' has no attribute 'llms' @slundberg
